File: apps/kabegame/src/services/updater.ts
//! 应用自动更新服务（前端镜像编排）。
//!
//! 状态归后端 `UpdaterService` 权威；本模块只负责：启动 hydrate + 订阅事件 +
//! 转发用户操作（手动检查 / 下载 / 取消 / 重启）。调度在后端，前端不再 setInterval。

import { invoke, listen, type UnlistenFn } from "@/api/rpc";
import { IS_WEB, IS_ANDROID } from "@kabegame/core/env";
import { kameMessage } from "@kabegame/core/utils/kameMessage";
import {
  useUpdaterStore,
  type DownloadProgress,
  type ReleaseInfo,
  type UpdaterPhase,
  type UpdaterState,
} from "@/stores/updater";

let unlistenState: UnlistenFn | null = null;
let unlistenProgress: UnlistenFn | null = null;
let unlistenError: UnlistenFn | null = null;

/** 桌面以外（web / android）全程 noop：后端命令在这些目标未注册。 */
function disabled(): boolean {
  return IS_WEB || IS_ANDROID;
}

/** App.vue onMounted 调一次：hydrate 当前状态 + 订阅三事件。 */
export async function init(): Promise<void> {
  if (disabled()) return;
  const store = useUpdaterStore();

  // 1) 启动 hydrate（与 organize 一致；刷新页面也靠它恢复状态）
  try {
    store.applyState(await invoke<UpdaterState>("get_updater_state"));
  } catch (e) {
    console.warn("[updater] get_updater_state failed:", e);
  }

  // 2) 订阅后端事件
  unlistenState = await listen<UpdaterState>("updater-state-change", (e) => {
    store.applyState(e.payload);
  });
  unlistenProgress = await listen<DownloadProgress>("update-download-progress", (e) => {
    store.applyProgress(e.payload);
  });
  unlistenError = await listen<{ message?: string }>("update-download-error", (e) => {
    const msg = e.payload?.message ?? "";
    store.setDownloadError(msg);
    if (msg) kameMessage.error(msg); // 龟酱播报错误
  });
}

export function dispose(): void {
  unlistenState?.();
  unlistenProgress?.();
  unlistenError?.();
  unlistenState = unlistenProgress = unlistenError = null;
}

/** 手动触发检查（Settings 按钮）。busy 时后端 no-op。返回最新快照的 phase。 */
export async function checkNow(): Promise<UpdaterPhase | null> {
  if (disabled()) return null;
  const store = useUpdaterStore();
  try {
    const snap = await invoke<UpdaterState>("check_for_updates");
    store.applyState(snap);
    return snap.phase;
  } catch (e) {
    console.warn("[updater] check_for_updates failed:", e);
    return null;
  }
}

/** 开始下载某 release（仅 updateAvailable 可进入；下载/检查中后端拒绝）。 */
export async function downloadAndStage(r: ReleaseInfo): Promise<void> {
  if (disabled()) return;
  await invoke("download_update", {
    tag: r.tag,
    assetUrl: r.assetUrl,
    assetName: r.assetName,
  });
}

/** 取消进行中的下载。 */
export async function cancelDownload(): Promise<void> {
  if (disabled()) return;
  await invoke("cancel_download");
}

/** 应用已下载更新并重启（restartable 下调用）。 */
export async function applyUpdateAndRestart(): Promise<void> {
  if (disabled()) return;
  await invoke("apply_update_and_restart");
}

File: src-tauri/kabegame/src/commands/updater.rs
//! Tauri 命令薄包装：应用自动更新。具体逻辑在 `crate::updater`。
//!
//! 状态归后端 `UpdaterService` 权威；前端启动 `get_updater_state` hydrate，
//! 之后靠 `updater-state-change` / `update-download-progress` / `update-download-error` 事件刷新。
//! `checking` 与 `downloading` 独占、不可重入。

use crate::updater::{self, UpdaterService, UpdaterState};

/// 返回当前更新状态快照（前端启动期 hydrate）。
#[tauri::command]
pub async fn get_updater_state() -> Result<UpdaterState, String> {
    Ok(UpdaterService::global().snapshot())
}

/// 手动检查更新；`checking`/`downloading` 期间 no-op，返回当前快照（不重入、不打断）。
#[tauri::command]
pub async fn check_for_updates() -> Result<UpdaterState, String> {
    UpdaterService::global().run_check().await;
    Ok(UpdaterService::global().snapshot())
}

/// 下载指定 release 的安装包到临时目录（仅可从 `updateAvailable` 进入）。
#[tauri::command]
pub async fn download_update(
    tag: String,
    asset_url: String,
    asset_name: String,
) -> Result<(), String> {
    updater::download_update(tag, asset_url, asset_name).await
}

/// 取消进行中的下载；仅 `downloading` 有效。返回是否真的发起取消。
#[tauri::command]
pub async fn cancel_download() -> Result<bool, String> {
    Ok(UpdaterService::global().cancel_download())
}

/// 应用已下载的更新并重启（restartable 下调用）。Linux 不支持。
#[tauri::command]
pub async fn apply_update_and_restart(app: tauri::AppHandle) -> Result<(), String> {
    let (_tag, path) = UpdaterService::global()
        .downloaded_package()
        .ok_or("没有已下载的更新包")?;
    updater::apply_update(&app, &path)
}
